Adding better error messages.

jeeyoungk opened the pull request:
* Adding more structured errors, to share structure between similar error types.
* Adding the type `ErrorList`, which can be used to return multiple parsing errors at the same time. This can detect things like multiple env vars not being in place.

ntindall commented on envvar/envvar.go:
```diff
@@ -93,9 +102,53 @@ type UnsetVariableError struct {
+	for _, err := range e.errors {
+		allErrors = append(allErrors, err.Error())
+	}
+	return fmt.Sprintf("envvar:\n%s", strings.Join(allErrors, "\n"))
```
I think this will just print out something like

```
envvar:
Err1
Err2
Err3
```

Should we just do 
```
envvar: err1
envvar: err2
envvar: err3
```

seems like this would avoid ambiguity in log messages.

jeeyoungk replied:
This behavior has been updated.

ntindall commented on envvar/envvar.go:
```diff
@@ -44,46 +44,55 @@ func Parse(v interface{}) error {
 	val := reflect.ValueOf(v)
 	if val.IsNil() {
-		return fmt.Errorf("envvar: Error in Parse: argument cannot be nil.")
+		return fmt.Errorf("envvar: Error in Parse: argument cannot be nil")
```
`InvalidArgumentError`? (it would be conventient to be able to just do a type assertion on the return value of `Parse` and have it always be an `ErrorList` (would also allow us to put more errors under the hood, potentially).

ntindall commented on envvar/envvar.go:
```diff
@@ -93,9 +102,53 @@ type UnsetVariableError struct {
 
+// InvalidVariableError is returned when a given env var cannot be parsed to
+// a given struct field.
+type InvalidVariableError struct {
```
so by exporting these we are saying it is ok to do a type assertion and inspect the contents. Is there a use case for that @jeeyoungk ? Perhaps if you wanted more business logic around the way to default environment variables (seems like this is antithetical to the idea of an environment variable).

jeeyoungk replied:
I would love to be able to test it (at least), and expose more structure if necessary. I think making it being a formal API contract would be good also.

ntindall commented on envvar/envvar.go:
```diff
@@ -93,9 +102,53 @@ type UnsetVariableError struct {
+// ErrorList is list of independent errors raised by Parse
+type ErrorList struct {
+	errors []error
+}
```
Should we do:

```go
var _ error = &ErrorList{}
```

(same for the other errors?) maybe there is no point in doing this since we are returning it as an `error` in `Parse`.

jeeyoungk replied:
i feel like it's obvious enough interface that we don't need to satisfy the interface. I think it's pretty obvious.

ntindall commented:
This fixes the second issue / enhancement identified here: https://github.com/plaid/go-envvar/issues/7

jeeyoungk commented:
Ok I updated.

I removed `getStructTag` as this is now implemented in the stdlib as `Lookup`. This allows us to remove duplicated code & reduce # of errors.

https://github.com/plaid/go-envvar/pull/8/files?w=1 gives much better view of the diff without the whitespace

ntindall commented on envvar/envvar_test.go:
```diff
@@ -141,6 +159,75 @@ func TestParseWithInvalidArgs(t *testing.T) {
+	if err == nil {
+		t.Errorf("Expected InvalidVariableError, but got nil error")
+	} else if _, ok := err.(InvalidVariableError); !ok {
+		t.Errorf("Expected InvalidVariableError, but got %s", err)
```
I think this will not scan into `t.Errorf` correctly @jeeyoungk 
we should have `"Expected InvalidVariableError, but got %T"`

ntindall commented on envvar/envvar_test.go:
```diff
@@ -141,6 +159,75 @@ func TestParseWithInvalidArgs(t *testing.T) {
+	value := reflect.ValueOf(xptr).Elem()
+	expectInvalidVariableError(t, setFieldVal(value, "name", "abc"))
+	if err := setFieldVal(value, "name", "42.3"); err != nil {
+		t.Errorf("Unexpected error on setFieldVal(): %s", err)
```
similarly `s/%s/%T/`

ntindall commented on envvar/envvar_test.go:
```diff
@@ -141,6 +159,75 @@ func TestParseWithInvalidArgs(t *testing.T) {
+	value := reflect.ValueOf(xptr).Elem()
+	expectInvalidVariableError(t, setFieldVal(value, "name", "not-bool"))
+	if err := setFieldVal(value, "name", "true"); err != nil {
+		t.Errorf("Unexpected error on setFieldVal(): %s", err)
```
I think we need to do `err.Error()` in order for `t.Errorf` to format correctly.

ntindall commented on envvar/envvar_test.go:
```diff
@@ -141,6 +159,75 @@ func TestParseWithInvalidArgs(t *testing.T) {
+	value := reflect.ValueOf(xptr).Elem()
+	expectInvalidVariableError(t, setFieldVal(value, "name", "-3"))
+	if err := setFieldVal(value, "name", "15"); err != nil {
+		t.Errorf("Unexpected error on setFieldVal(): %s", err)
```
`s/%s/%T/`

ntindall commented on envvar/envvar_test.go:
```diff
@@ -141,6 +159,75 @@ func TestParseWithInvalidArgs(t *testing.T) {
+	value := reflect.ValueOf(xptr).Elem()
+	expectInvalidVariableError(t, setFieldVal(value, "name", "abc"))
+	if err := setFieldVal(value, "name", "15"); err != nil {
+		t.Errorf("Unexpected error on setFieldVal(): %s", err)
```
`s/%s/%T/`

ntindall commented on envvar/envvar.go:
```diff
@@ -93,45 +97,51 @@ type UnsetVariableError struct {
+type InvalidVariableError struct {
+	VarName  string
+	VarValue string
+	parent   error // optional
```
I wonder what the implications of not exporting this to the caller are.

I think it is OK.

ntindall commented on envvar/envvar.go:
```diff
@@ -39,51 +37,57 @@ func Parse(v interface{}) error {
+
+func parse(field reflect.StructField, fieldVal reflect.Value) error {
+	varName := field.Name
+	customName := field.Tag.Get("envvar")
```
i didn't realize the `customName` functionality was _optional_... e.g. if you name your variables things like `MY_ENVIRONMENT_VARIABLE` it just works?

should we document this in the `README`?

jeeyoungk replied:
technically optional but i think it's difficult to use as it goes against the go style guide.

also you cant have lowercase names as it would be hopelessly private as per our god Rob Pike directed to be so.

ntindall commented:
@jeeyoungk after we resolve the comments, lets merge this so that @albrow can rebase https://github.com/plaid/go-envvar/pull/9

jeeyoungk commented:
Updated tests & readme.

...who can merge this?

ntindall commented:
thanks @jeeyoungk !
